### PR TITLE
Chunk the image horizontally (for landscape images) or vertically (for portrait images) and evaluate faces in each chunk

### PR DESCRIPTION
**Pull request by JPPhoto**
When an image fails initial finding, we now break it into square chunks and attempt the face finding in each chunk.

If all attempts fail, return the whole image as the face-found image so as not to throw an exception.

### CODE REVIEW

File: faceoff.py [hunk 1]
```diff
@@ -29,20 +66,21 @@ class FaceOffOutput(BaseInvocationOutput):
...
 
-@invocation("face_off", title="FaceOff", tags=["image", "faceoff", "face", "mask"], category="image")
+@invocation("face_off", title="FaceOff", tags=["image", "faceoff", "face", "mask"], category="image", version="1.0.0")
 class FaceOffInvocation(BaseInvocation):
     """bound, extract, and mask a face from an image using MediaPipe detection"""
 
     image:               ImageField  = InputField(description="Image for face detection")
     face_id:             int = InputField(default=0, description="0 for first detected face, single digit for one specific. Multiple faces not supported. Find a face's ID with FaceIdentifier node.")
-    faces:               int = InputField(default=4, description="Maximum number of faces to detect")
+    faces:               int = InputField(default=99, description="Maximum number of faces to detect")
```
Comment (ymgenesis) [inline faceoff.py]: Maybe we can just use 10 as a default. It's probably not likely anyone will need more than that and be caught unawares.
```suggestion
    faces:               int = InputField(default=10, description="Maximum number of faces to detect")
```

File: faceoff.py [hunk 2]
```diff
@@ -63,70 +101,115 @@ def generate_face_box_mask(self, pil_image):
...
+                # Give up
+                print(f"No face detected in chunked input image. Passing through original image.")
+                raise ValueError("No face detected in input image.")
+
+        result = cleanup_faces_list(result)
+
+        face_id = self.face_id - 1 if self.face_id > 0 else self.face_id
+
+        if face_id > len(result) or face_id < 0:
+            raise ValueError("Selected face ID is outside of the number of faces detected.")
```
Comment (ymgenesis) [inline faceoff.py]:
```suggestion

```
This one isn't really raised, can probably remove it and just have a general print statement in the invoke class before passing through the original image.

**Review by ymgenesis — COMMENTED** [review]
In general it's kind of interesting to pass through the original image when failing... but I guess it's better than stopping the entire graph for now.

I think more testing probably needs to be done with larger portrait images with multiple faces. As mentioned, I'll work on faceidentifier with this implementation and if I can get them to work together we can implement it with faceidentifier and facemask as well.

File: faceoff.py (hunk 2 above)
Comment (JPPhoto) [reply faceoff.py]: I think it's raised if you ask for face 3 and there are only 2, for example. If we don't handle it here, you get an out of bounds index on the array in the next line.

File: faceoff.py [hunk 3]
```diff
@@ -77,10 +77,11 @@ class FaceOffInvocation(BaseInvocation):
     x_offset:            float = InputField(default=0.0, description="X-axis offset of the mask")
     y_offset:            float = InputField(default=0.0, description="Y-axis offset of the mask")
     padding:             int = InputField(default=0, description="All-axis padding around the mask in pixels")
+    chunk:               bool = InputField(default=False, description="Whether to bypass full image face detection and default to image chunking. Chunking will occur if no faces are found in the full image.")
 
     def generate_face_box_mask(self, pil_image, chunk_x_offset=0, chunk_y_offset=0):
         result = []
-
+
```
Comment (ymgenesis) [inline faceoff.py]: -_-

**Pull request merged**